### PR TITLE
model: eigen decomposition /  linear equation solver / SVD / inverse

### PR DESCRIPTION
**Pull request by sugarspectre**

### CODE REVIEW

File: src/preprocessing/encoder.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
+export class OneHotEncoder {
```
Comment (FeelyChau) [inline src/preprocessing/encoder.ts:2]: Maybe we should remove this until we implement it.

File: test/node/linalg/eigen.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+    
+  it('spectrum decomposition', () => {
+    const [ d, q ] = eigenSolve(matrix);
+    const di = tf.diag(d);
+    const recovM = tf.matMul(tf.matMul(tf.transpose(q), di), q);
+    const dm = tf.matMul(tf.matMul(q, matrix), tf.transpose(q));
+    const equal = tensorEqual(recovM, matrix, 1e-3);
+    assert.isTrue(equal);
+  })
+})
```
Comment (FeelyChau) [inline test/node/linalg/eigen.ts:26]: EOF.

File: test/node/linalg/linsolve.ts [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+const v = tensor([0, 8, 10]);
+
+describe('linSolver', () => {
+  it('solve matrix', () => {
+    const solve = linSolveQR(matrix, v);
+    const predv = squeeze(matMul(matrix, reshape(solve, [-1, 1])));
+    const equal = tensorEqual(v, predv, 1e-3);
+    assert.isTrue(equal);
+  })
+})
```
Comment (FeelyChau) [inline test/node/linalg/linsolve.ts:21]: EOF.

File: src/linalg/eigen.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    //preQ = q;
+    x = matMul(r, q);
+    [ q, r ] = linalg.qr(x);
+    qn = matMul(qn, q);
+    xTr = linalg.bandPart(x, 0, 0);
+    const maxDis = max(abs(sub(x, xTr))).arraySync();
+    if ( maxDis < tol) {
+      break;
+    }
+    //console.log(maxDis);
```
Comment (FeelyChau) [inline src/linalg/eigen.ts:39]: Useless line.

File: src/linalg/eigen.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+ * @param matrix square matrix of shape (n, n)
+   @param tol tolerence, default to 1e-4
+   @param maxIter max iteration time, default to 200
+ */
+export const solveEigenValues = (matrix: Tensor, tol = 1e-4, maxIter = 200): Tensor => {
+  let [ q, r ] = linalg.qr(matrix);
+  let x = matrix;
+  const n = matrix.shape[0];
+  let xTr = linalg.bandPart(x, 0, 0);
+  //let preQ = q;
```
Comment (FeelyChau) [inline src/linalg/eigen.ts:26]: Useless line.

File: src/linalg/eigen.ts [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  let qn = q;
+
+  for (let i = 0; i < maxIter; i++) {
+    //preQ = q;
```
Comment (FeelyChau) [inline src/linalg/eigen.ts:30]: Useless line.

**Review by FeelyChau — CHANGES_REQUESTED** [review]

**Comment by FeelyChau** [comment]
@rickycao-qy May I ask you to review the part of the algorithm?

File: test/node/linalg/svd.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+const singularValues = tf.tensor([18.54, 5.01, 1.83]);
+
+describe('SVDSolver', () => {
+    
+    it('svd decomposition', () => {
+      const [ u, d, v ] = svd(matrix);
+      const equal = tensorEqual(d, singularValues, 1e-2);
+      assert.isTrue(equal);
+    })
+  });
```
Comment (yorkie) [inline test/node/linalg/svd.ts:22]: Wrong indentions.

File: test/node/linalg/linsolve.ts [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+])
+const v = tensor([0, 8, 10]);
+
+describe('linSolver', () => {
+  it('solve matrix', () => {
+    const solve = linSolveQR(matrix, v);
+    const predv = squeeze(matMul(matrix, reshape(solve, [-1, 1])));
+    const equal = tensorEqual(v, predv, 1e-3);
+    assert.isTrue(equal);
+  })
```
Comment (yorkie) [inline test/node/linalg/linsolve.ts:20]:
```suggestion
  });
```

File: test/node/linalg/inverse.ts [hunk 9]
```diff
@@ -0,0 +1,23 @@
...
+describe('EigenSolver', () => {
+    
+  it('spectrum decomposition', () => {
+    const invM = inverse(matrix);
+    const iM = tf.matMul(matrix, invM);
+    const isIdMatrix = tensorEqual(iM, tf.eye(iM.shape[0]), 1e-2);
+    assert.isTrue(isIdMatrix);
+  });
+  
+});
```
Comment (yorkie) [inline test/node/linalg/inverse.ts:23]: EOF

File: test/node/linalg/inverse.ts [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+import * as tf from '@tensorflow/tfjs-core';
+import { tensorEqual } from '../../../src/linalg/utils';
+import { assert } from 'chai';
+import 'mocha';
+const matrix = tf.tensor2d([
+    [ 1, 2, 5, 10 ],
+    [ 2, 6, 7, 5 ],
+    [ 5, 7, 9, 6 ],
+    [ 10, 5, 6, 7]
+  ]);
```
Comment (yorkie) [inline test/node/linalg/inverse.ts:12]: Ditto.

File: test/node/linalg/eigen.ts [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+
+describe('EigenSolver', () => {
+    
+  it('spectrum decomposition', () => {
+    const [ d, q ] = eigenSolve(matrix);
+    const di = tf.diag(d);
+    const recovM = tf.matMul(tf.matMul(q, di), tf.transpose(q));
+    const equal = tensorEqual(recovM, matrix, 1e-3);
+    assert.isTrue(equal);
+  })
```
Comment (yorkie) [inline test/node/linalg/eigen.ts:23]:
```suggestion
  });
```

File: src/linalg/utils.ts [hunk 12]
```diff
@@ -0,0 +1,91 @@
...
+ * @param n number of columns
+ */
+export const fillDiag = (values: Tensor | number[], m: number, n: number): Tensor => {
+  const svSize = m > n ? n : m;
+  const valuesData = values instanceof Tensor ? values.dataSync() : values;
+  const vSize = valuesData.length;
+  const diagData = new Array(m).fill(0).map( () => new Array(n).fill(0));
+  for (let i = 0; i < svSize; i++) {
+    if (i < vSize) {
+      diagData[i][i] = Number(slice(valuesData, i, 1).dataSync());
```
Comment (yorkie) [inline src/linalg/utils.ts:87]:
```suggestion
      diagData[i][i] = parseInt(slice(valuesData, i, 1).dataSync());
```

File: src/linalg/utils.ts [hunk 13]
```diff
@@ -0,0 +1,91 @@
...
+ * Check that two tensors are equal to within some additive tolerance.
+ * @param tensor1
+ * @param tensor2
+ * @param
+ */
+export const tensorEqual = (tensor1: Tensor, tensor2: Tensor, tol = 0): boolean => {
+  if (!shapeEqual(tensor1, tensor2)) {
+    throw new Error('tensor1 and tensor2 not of same shape');
+  }
+  const isEqual = lessEqual(max(abs(sub(tensor1, tensor2))), tol).dataSync();
```
Comment (yorkie) [inline src/linalg/utils.ts:66]:
```suggestion
  return lessEqual(max(abs(sub(tensor1, tensor2))), tol).dataSync();
```

File: src/linalg/eigen.ts [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+  set of eigenvalues.  Starting with the matrix M = M0, we iterate:
+    M0 = Q0 R0, M1 = R0 Q0;
+    M1 = Q1 R1, M2 = R1 Q1;
+    M2 = Q2 R2, M3 = R2 Q2;
+    ...
+  For a general matrix with a full set of eigenvalues, this sequence will
+  converge to an upper diagonal matrix:
+    Mi -> upper diagonal matrix
+  The diagonal entries of this matrix are the eigenvalues of M.
+  Qn is the eigenvector of M.
```
Comment (yorkie) [inline src/linalg/eigen.ts:16]: The above lines must prepend "*" to keep the alignments.

File: src/linalg/eigen.ts [hunk 15]
```diff
@@ -0,0 +1,123 @@
...
+
+  for (let i = 0; i < maxIter; i++) {
+    prevX = x;
+    x = matMul(r, q);
+    [ q, r ] = linalg.qr(x);
+    qn = matMul(qn, q);
+    xTr = linalg.bandPart(x, 0, 0);
+    prevTr = linalg.bandPart(prevX, 0, 0);
+    const maxDis = max(abs(sub(prevTr, xTr))).arraySync();
+    if ( maxDis < tol) {
```
Comment (yorkie) [inline src/linalg/eigen.ts:38]:
```suggestion
    if (maxDis < tol) {
```

File: src/linalg/eigen.ts [hunk 16]
```diff
@@ -0,0 +1,123 @@
...
+      break;
+    }
+  }
+  x = matMul(r, q);
+  const eigenValues = [];
+  for (let i = 0; i < n; i++) {
+    eigenValues.push(slice(x, [ i, i ], [ 1, 1 ]));
+  }
+  const eigenValuesTensor = squeeze(stack(eigenValues));
```
Comment (yorkie) [inline src/linalg/eigen.ts:47]:
```suggestion
  return squeeze(stack(eigenValues));
```

No necessary to declaration & assignment ops.

File: src/linalg/eigen.ts [hunk 17]
```diff
@@ -0,0 +1,123 @@
...
+    ...
+  This algorithm will converge to the eigenvector associated with the eigenvalue
+  closest to lambda.
+ * @param matrix matrix
+ * @param eigenValue eigen value
+ * @param tol tolerance, default to 1e-4
+ * @param maxIter max iteration time, default to 200
+ */
+export const eigenBackSolve = (matrix: Tensor, eigenValue: number, tol = 1e-4, maxIter = 200): Tensor => {
+  const n = matrix.shape[0];
```
Comment (yorkie) [inline src/linalg/eigen.ts:70]: What does the variable `n` mean? Do we have a more meaningful name?

File: src/linalg/eigen.ts [hunk 18]
```diff
@@ -0,0 +1,123 @@
...
+  for (let i = 0; i < maxIter; i++) {
+    previous = current;
+    current = linSolveQR(mi, previous);
+    current = tensorNormalize(current);
+    /**
+     * We reverse the sign of the vector if the first entry is not positive.
+     * Often the algorithm will oscilate between a vector and its negative
+     * after convergence.
+     */
+    const pivot = Number(slice(current, 0, 1).dataSync());
```
Comment (yorkie) [inline src/linalg/eigen.ts:86]:
```suggestion
    const pivot = parseInt(slice(current, 0, 1).dataSync());
```

File: src/linalg/eigen.ts [hunk 19]
```diff
@@ -0,0 +1,123 @@
...
+    if (tensorEqual(current, previous, tol)) {
+      break;
+    }
+  }
+  return current;
+};
+
+/* Solve for the eigenvectors of a matrix M once the eigenvalues are known
+   using inverse iteration.
+*/
```
Comment (yorkie) [inline src/linalg/eigen.ts:99]:
```suggestion
/* 
 * Solve for the eigenvectors of a matrix M once the eigenvalues are known
 * using inverse iteration.
 */
```

File: src/linalg/eigen.ts [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+export const solveEigenVectors = (matrix: Tensor, eigenValues: Tensor, tol = 1e-4, maxIter = 200): Tensor => {
+  const nEv = eigenValues.shape[0];
```
Comment (yorkie) [inline src/linalg/eigen.ts:101]: Good name.

File: src/linalg/eigen.ts [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+  const eigenVectors = [];
+  for (let i = 0; i < nEv; i++) {
+    const eigenValue = Number(slice(eigenValues, i, 1).dataSync());
```
Comment (yorkie) [inline src/linalg/eigen.ts:104]: Ditto

File: src/linalg/inverse.ts [hunk 22]
```diff
@@ -0,0 +1,27 @@
...
+    const invertable = Boolean(greater(minEigen, 1e-4));
+    if (!invertable) {
+      throw new Error('Singlular matrix error');
+    } else {
+      const inverseEigenValues = divNoNan(1, eigenValues);
+      const inverseM = matMul(matMul(eigenVectors, diag(inverseEigenValues)), transpose(eigenVectors));
+      return inverseM;
+    }
+  } else {
+    throw new Error('Singular matrix error');
```
Comment (yorkie) [inline src/linalg/inverse.ts:25]:
```suggestion
    throw new TypeError('Singular matrix error');
```

File: src/linalg/inverse.ts [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+export const inverse = (matrix: Tensor): Tensor => {
+  if (isSquareMatrix) {
+    const [ eigenValues, eigenVectors ] = eigenSolve(matrix);
+    const minEigen = min(abs(eigenValues));
+    const invertable = Boolean(greater(minEigen, 1e-4));
+    if (!invertable) {
+      throw new Error('Singlular matrix error');
+    } else {
+      const inverseEigenValues = divNoNan(1, eigenValues);
+      const inverseM = matMul(matMul(eigenVectors, diag(inverseEigenValues)), transpose(eigenVectors));
```
Comment (yorkie) [inline src/linalg/inverse.ts:21]:
```suggestion
      return matMul(matMul(eigenVectors, diag(inverseEigenValues)), transpose(eigenVectors));
```

File: src/linalg/linsolve.ts [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+   then solving the resulting equation for x_{l-1}.  Continuing in this way
+   solves the entire system.
+ * @param r upper triangle matrix
+ * @param v target value
+ */
+export const linSolveUpperTriangle = (r: Tensor, v: Tensor): Tensor => {
+  const nEq = v.shape[0];
+  const solution = new Array(nEq).fill(0);
+  let backSubsitute: number;
+  // eslint-disable-next-line for-direction
```
Comment (yorkie) [inline src/linalg/linsolve.ts:22]: Add a comment to describe why do we disable the lint rule?

File: src/linalg/linsolve.ts [hunk 25]
```diff
@@ -0,0 +1,60 @@
...
+   reduces the problem to solving a linear equation Rx = y for an upper
+   triangular matrix R.
+*/
+export const linSolveFromQR = (q: Tensor, r: Tensor, v: Tensor): Tensor => {
+  const rhs = squeeze(matMul(transpose(q), reshape(v, [ -1, 1 ])));
+  const solution = linSolveUpperTriangle(r, rhs);
+  return solution;
+};
+
+export const linSolveQR = (matrix: Tensor, v: Tensor): Tensor => {
```
Comment (yorkie) [inline src/linalg/linsolve.ts:56]: Add a comment for the function.

File: src/linalg/utils.ts [hunk 26]
```diff
@@ -0,0 +1,91 @@
+import { Tensor, norm, div, max, sub, abs, lessEqual, slice, tensor } from "@tensorflow/tfjs-core";
```
Comment (yorkie) [inline src/linalg/utils.ts:1]:
```suggestion
import { Tensor, norm, div, max, sub, abs, lessEqual, slice, tensor } from '@tensorflow/tfjs-core';
```
Must use one way for string literal.

File: src/linalg/svd.ts [hunk 27]
```diff
@@ -0,0 +1,22 @@
+import { eigenSolve } from "./eigen";
+import { matMul, Tensor, transpose, sqrt } from "@tensorflow/tfjs-core";
```
Comment (yorkie) [inline src/linalg/svd.ts:2]: Ditto.
```suggestion
import { matMul, Tensor, transpose, sqrt } from '@tensorflow/tfjs-core';
```

**Review by yorkie — CHANGES_REQUESTED** [review]

File: src/linalg/linsolve.ts [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+  const nEq = v.shape[0];
+  const solution = new Array(nEq).fill(0);
+  let backSubsitute: number;
+  // eslint-disable-next-line for-direction
+  for (let i = nEq - 1; i >= 0; i--) {
+    backSubsitute = 0;
+    if (i < nEq - 1) {
+      for (let j = i + 1; j <= nEq - 1; j++) {
+        const l = solution[j];
+        const w = Number(slice(r, [ i, j ], [ 1, 1 ]).dataSync());
```
Comment (rickycao-qy) [inline src/linalg/linsolve.ts:28]: There are lots of dataSync method which may cause performance issues. Suggest to change them to ```await tensor.todata()```

**Review by rickycao-qy — CHANGES_REQUESTED** [review]

File: src/linalg/utils.ts (hunk 12 above)
Comment (sugarspectre) [reply src/linalg/utils.ts:82]: parseInt will not work in this case, since the return type of dataSync() method is not determined

**Review by FeelyChau — APPROVED** [review]
LGTM.

**Pull request merged**